### PR TITLE
Migrating to mongo client 4.x

### LINKED ISSUE #2338: Upgrade to Mongo 4 java driver
Currently the Mongo DB instrumentation references legacy classes in the Mongo 3 java driver. Spring Boot requires the Mongo 4 driver starting with 2.3.0.RELEASE. 

There are 2 Mongo listeners that provide metrics:
* [MongoMetricsCommandListener](https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsCommandListener.java#L18)
* [MongoMetricsConnectionPoolListener](https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java#L18)

They both reference `com.mongo.MongoClient` (only for a Javadoc `@link` tag) which still exists in the Mongo 4 java driver, but its in the `mongodb-driver-legacy` lib. The connection pool listener, however, actually references event classes that have been removed in Mongo 4 driver. 

Suggested plan:

* Upgrade the Mongo dependency to [mongodb-driver-sync](https://mvnrepository.com/artifact/org.mongodb/mongodb-driver-sync/4.0.5)
* Replace the `@link` tag w/ `{@code }` therefore removing the dependency on any MongoClient (its not used anyways)
* Replace the usage of the deleted event classes in the connection pool listener w/ their Mongo 4 equivalent events

Concerns:
* This will force Mongo 4 on consumers of the instrumentation. It would have to be in a minor version release. There may 
be a way to use an lowest common denominator of Mongo 3 and 4 for the connection pool listener events so as to be usable by both Mongo 3 and 4.

### PR DESCRIPTION
**Pull request by jonatan-ivanov**
Resolves #2338

### CODE REVIEW

**Review by bono007 — COMMENTED** [review]
@jonatan-ivanov I added a few comments. Thx for doing this!

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java [hunk 1]
```diff
@@ -79,6 +79,11 @@ public void connectionPoolClosed(ConnectionPoolClosedEvent event) {
         waitQueueSize.remove(serverId);
     }
 
+//    @Override
+//    public void connectionPoolCleared(ConnectionPoolClearedEvent event) {
+//        // TODO: Do we need to implement this?
```
Comment (shakuzen) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java:84]: Is the ConnectionClosedEvent called for each connection when this happens? If so, perhaps we don't need to do anything with this event. The pool is still open and usable after it is cleared, right?

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java [hunk 2]
```diff
@@ -68,7 +73,8 @@ public void clusterOpening(ClusterOpeningEvent event) {
 
         assertThat(registry.get("mongodb.driver.pool.size").tags(tags).gauge().value()).isEqualTo(2);
         assertThat(registry.get("mongodb.driver.pool.checkedout").gauge().value()).isZero();
-        assertThat(registry.get("mongodb.driver.pool.waitqueuesize").gauge().value()).isZero();
+        // TODO: waitQueueEntered and waitQueueExited were removed, how can we provide mongodb.driver.pool.waitqueuesize?
```
Comment (shakuzen) [inline micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java:76]: We should check why it was removed. If it isn't considered useful, then we should just note its removal in our release notes and reference Mongo changelogs/release notes/documentation explaining its removal.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsCommandListener.java [hunk 3]
```diff
@@ -15,7 +15,7 @@
  */
 package io.micrometer.core.instrument.binder.mongodb;
 
-import com.mongodb.MongoClient;
+import com.mongodb.client.MongoClient;
 import com.mongodb.event.*;
```
[the remark this answers is not in the document]
Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsCommandListener.java:19]: The Micrometer style I think has best been described as the IntelliJ default (which will use wildcard imports if there are above a certain number of classes imported from the same package). We don't strictly enforce import style with Checkstyle checks, though.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java [hunk 4]
```diff
@@ -79,6 +79,11 @@ public void connectionPoolClosed(ConnectionPoolClosedEvent event) {
         waitQueueSize.remove(serverId);
     }
 
+//    @Override
```
[the remark this answers is not in the document]
Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java:82]: What metrics would you suggest we make related to these events?

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (shakuzen) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java:76]: Do you happen to know anything from the Mongo team we could reference in our release notes? I asked about it in https://github.com/micrometer-metrics/micrometer/pull/2493#discussion_r590964124 before I saw your review comments.

File: dependencies.gradle [hunk 5]
```diff
@@ -70,7 +70,7 @@ def VERSIONS = [
         'org.latencyutils:LatencyUtils:latest.release',
         'org.mockito:mockito-core:latest.release',
         'org.mockito:mockito-inline:latest.release',
-        'org.mongodb:mongo-java-driver:latest.release',
+        'org.mongodb:mongodb-driver-sync:latest.release',
```
Comment (shakuzen) [inline dependencies.gradle:73]: This reminded me: I wonder what it would take to support instrumenting other drivers than the sync one. It's not something that needs to be covered in this pull request, but I just thought I'd mention it.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (bono007) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java:82]: I'm not completely sure - i have not thought too much about it. I would reword my comment slightly...

> I am in favor of implementing these.

to "I am in favor of investigating these events a bit and seeing if it makes sense to add them".

It could also be done on a follow on PR once the base 3.x->4.x change is in.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java (hunk 1 above)
Comment (bono007) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java:84]: Good questions @shakuzen . Seeing this plus your other comment about the other new event types I am in favor of moving these "new event support" to a follow on issue. This ticket is really to move from 3.x -> 4.x and this current proposal puts us in parity w/ 3.x sans "waitqueuesize".

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bono007) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java:76]: Not from the Mongo team per-se. However, [this deprecation list](https://mongodb.github.io/mongo-java-driver/3.12/javadoc/deprecated-list.html) (posted in 3.12) was super helpful when we migrated our apps to 4.x.

And [drilling down](https://mongodb.github.io/mongo-java-driver/3.12/javadoc/com/mongodb/event/ConnectionPoolWaitQueueEnteredEvent.html) directly into the event in question you see this loud warning:
| ⚠️  Deprecated. In the next major release, this event will no longer be published |
| --- |

Comment (bono007) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java:76]: Here is [the motivation](https://jira.mongodb.org/browse/JAVA-3460) for the removal.

[the remark this answers is not in the document]
Comment (shakuzen) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java:76]: I see. We can certainly reference that. I was hoping for any reasoning why it was deprecated and removed, but lacking that, we can use the deprecation list you linked. As a user, did you ever use the wait queue metrics? Is there removal an issue for you or do you think it might be for other users?

**Comment by bono007** [comment]
I linked the actual motivation for the depreacation in my other comment.

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsCommandListenerTest.java [hunk 6]
```diff
@@ -51,15 +53,16 @@
     void setup() {
         registry = new SimpleMeterRegistry();
         clusterId = new AtomicReference<>();
-        MongoClientOptions options = MongoClientOptions.builder()
+        MongoClientSettings settings = MongoClientSettings.builder()
                 .addCommandListener(new MongoMetricsCommandListener(registry))
-                .addClusterListener(new ClusterListenerAdapter() {
+                .applyToClusterSettings(builder -> builder.hosts(Lists.newArrayList(new ServerAddress(HOST, port))))
```
[the remark this answers is not in the document]
Comment (jonatan-ivanov) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsCommandListenerTest.java:58]: Bad habits. :)
I used, Java higher versions of java a lot and I typed `Lists.`, then I realized: yepp, this is Java 8; but the AssertJ import was there which short circuited my brain just to use it. Good catch, thanks.

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonatan-ivanov) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java:76]: That JIRA issue is in connection with what I found: https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#why-are-waitqueuesize-and-waitqueuemultiple-deprecated

I'm going to ask the mongo community (`ConcurrentPool` has public `getCount`, `getInUseCount`, `getAvailableCount` but its interface does not and getting a reference could also be tricky.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java [hunk 7]
```diff
@@ -79,6 +79,11 @@ public void connectionPoolClosed(ConnectionPoolClosedEvent event) {
         waitQueueSize.remove(serverId);
```
[the remark this answers is not in the document]
Comment (jonatan-ivanov) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java:79]: Yepp, and the `waitQueueSize` map too. This is one of the reasons, why this PR is in draft.

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonatan-ivanov) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java:76]: Except these are not available on master. :)
I asked this on the mongo forum, if you are interested: https://developer.mongodb.com/community/forums/t/java-monitoring-waitqueue-size/98845

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java (hunk 1 above)
Comment (jonatan-ivanov) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java:80]: It is called when the pool is [invalidated](https://github.com/mongodb/mongo-java-driver/blob/c726328e75252c776857871e9c887a5e4ef12cc5/driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java#L253). This happens in error scenarios but nothing particular happens with the pool in this case. If the connection happens to be closed after this, as @shakuzen said a `ConnectionClosedEvent` will be emitted so we seems to be ok not implementing this.

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java (hunk 2 above)
Comment (jonatan-ivanov) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java:76]: I'm marking this resolved in favor of the other thread above, see my comments there: https://github.com/micrometer-metrics/micrometer/pull/2493#discussion_r593472189

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonatan-ivanov) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java]: Yepp, I saw them I'm just not sure if we should register any meter for them.
Maybe counting the `connectionCheckOutFailed` could be useful or a new counter on the `connectionCheckedOut`-`connectionCheckOutFailed` axis with different tags.

File: dependencies.gradle (hunk 5 above)
Comment (jonatan-ivanov) [reply dependencies.gradle:73]: https://github.com/micrometer-metrics/micrometer/issues/2498

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (bono007) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java]: I am in favor of splitting the "new" metrics off into a separate ticket. What do you think @jonatan-ivanov ?

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java (hunk 1 above)
Comment (bono007) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java]: > If the connection happens to be closed

Does it get closed after? I mean, that makes sense that it would, I just did not see it when glancing through the code (I did not pull down locally though either). Either way though, I think its fine to omit it now. If someone wants it they will file an issue.

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bono007) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java:76]: Between the deprecated list, release notes, and this spec outlining why the waitQueue metrics are deprecated it feels like the waitQueue metrics can be omitted.

[the remark this answers is not in the document]
Comment (bono007) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java:76]: Between the deprecated list, release notes, and this spec outlining why the waitQueue metrics are deprecated it feels like the waitQueue metrics can be omitted.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonatan-ivanov) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java]: 👍🏼 Feel free to create an issue. If you also want to PR it just let me know, I can assign it to you.

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bono007) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java]: I meant to ask this earlier. How are docs handled in Micrometer? Do we need to open a "documentation ticket" for the removal of waitqueue metric?

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (bono007) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java]: @jonatan-ivanov #2499

Thank you for the offer to assign it to me but I am not sure when I would get to it as I have my hands in a few cookie jars at the same time currently. :)

[the remark this answers is not in the document]
Comment (bono007) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java]: @jonatan-ivanov #2499

Thank you for the offer to assign it to me but I am not sure when I would get to it as I have my hands in a few cookie jars at the same time currently. :)

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonatan-ivanov) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java]: Depending on the outcome of the thread in the mongo forum, this can go to javadoc/release-notes/[micrometer-docs](https://github.com/micrometer-metrics/micrometer-docs).

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java (hunk 1 above)
Comment (jonatan-ivanov) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java]: This happens in error scenarios (e.g.: not able to connect) and what happens depends on when do you get the error:

1. It happens before the connection checked out:
This means the connection was not in use, there is a retry but since we din't even get the checked out event, we don't care, eventually there is a timeout and the connection will be closed
1. It happens after the connection checked out: 
This means that the connection was in use, there will be a check in and then a close which is good since it takes care about the events we need

Check out the `ConcurrentPool` (`release`, `prune`, `close`) and the maintenance task in `DefaultConnectionPool` they orchestrate this logic.

**Review by shakuzen — APPROVED** [review]

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bono007) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java]: Good call in posting on that forum @jonatan-ivanov - looks like we got some direction. It does sound like this will give roughly the same as waitqueue. I understand metrics are APIs and its nice to not break consumers. However, I feel like we should deprecate that metric as Mongo does not emit it anymore. What are your thoughts Jonatan and @shakuzen ? 

Also, what is that process like for Micrometer metrics?

[the remark this answers is not in the document]
Comment (shakuzen) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListenerTest.java]: I don't think any of the metrics backends we support have a concept of a deprecated metric (though it is an interesting idea), so unfortunately we're limited by what we can do to inform users. Release notes are typically where we would call out a removal. I agree we should consider whether it is an important metric for users or not. It seems the Mongo team thinks it isn't something to be tracked, judging from their removal of it. If there are Micrometer and MongoDB users out there that find it useful, we'd like to hear about that. We could drop an ask in the release notes and remove it in the next minor version if we don't hear people are using it.
